sql/expression/function/logarithm.go

```diff
@@ -0,0 +1,223 @@
+package function
+
+import (
+	"gopkg.in/src-d/go-mysql-server.v0/sql/expression"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+	"fmt"
+	"gopkg.in/src-d/go-errors.v1"
+	"math"
+	"reflect"
+)
+
+// ErrInvalidArgumentForLogarithm is returned when an invalid argument value is passed to a
+// logarithm function
+var ErrInvalidArgumentForLogarithm = errors.NewKind("invalid argument value for logarithm: %v")
+
+// LogBaseMaker returns LogBase creator functions with a specific base.
+func LogBaseMaker(base float64) func(e sql.Expression) sql.Expression {
+	b := base
+	return func(e sql.Expression) sql.Expression {
+		return NewLogBase(b, e)
+	}
+}
+
+// LogBase is a function that returns the logarithm of a value with a specific base.
+type LogBase struct {
+	expression.UnaryExpression
+	base float64
+}
+
+// NewLogBase creates a new LogBase expression.
+func NewLogBase(base float64, e sql.Expression) sql.Expression {
+	return &LogBase{UnaryExpression: expression.UnaryExpression{Child: e}, base: base}
+}
+
+func (l *LogBase) String() string {
+	switch l.base {
+	case float64(math.E):
+		return fmt.Sprintf("ln(%s)", l.Child)
+	case float64(10):
+		return fmt.Sprintf("log10(%s)", l.Child)
+	case float64(2):
+		return fmt.Sprintf("log2(%s)", l.Child)
+	default:
+		return fmt.Sprintf("log(%v, %s)", l.base, l.Child)
+	}
+}
+
+// TransformUp implements the Expression interface.
+func (l *LogBase) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	child, err := l.Child.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+	return f(NewLogBase(l.base, child))
+}
+
+// Type returns the resultant type of the function.
+func (l *LogBase) Type() sql.Type {
+	return sql.Float64
+}
+
+// IsNullable implements the sql.Expression interface.
+func (l *LogBase) IsNullable() bool {
+	return l.Child.IsNullable()
```

Comment:
IMO Logarithm can be NULL in ranges where function y=log b(x) doesn't exist:
- b or x is NULL
- base is <= 0 or base == 1
- x <= 0

Reply:
OK I will add the check here also

Reply:
I think it may work.